RavenDB-5042

ppekrol opened the pull request.

ppekrol committed:
RavenDB-5042 generating configuration entry metadata

ppekrol committed:
RavenDB-5042 properly calculating default values when they are set in ctor, added is secured information, changed model a bit to be more extensible

ppekrol committed:
RavenDB-5042 added indication that default value is set dynamically

ppekrol committed:
RavenDB-5042 marking all 'exec' configurations as secured

ppekrol committed:
RavenDB-5042 cleanup

ppekrol committed:
RavenDB-5042 added category

ayende commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -6,16 +6,17 @@
         [Description("A command or executable to run which will provide machine cpu usage and total machine cores to standard output. If specified, RavenDB will use this information for monitoring cpu usage.")]
         [DefaultValue(null)]
-        [ConfigurationEntry("Monitoring.Cpu.Exec", ConfigurationEntryScope.ServerWideOnly)]
+        [ConfigurationEntry("Monitoring.Cpu.Exec", ConfigurationEntryScope.ServerWideOnly, isSecured: true)]
```
Why is this secured?

ppekrol replied:
I've secured all 'exec' values

ayende commented on src/Raven.Server/Config/ConfigurationEntryValue.cs:
```diff
@@ -0,0 +1,120 @@
+                ServerValues[key] = new ConfigurationEntrySingleValue
+                {
+                    Value = hasAccess ? value : null,
+                    HasAccess = hasAccess
```
Where is this being used?

ppekrol replied:
It is for the UI, so we will be able to write there that the value is set, but user do not have access to it (e.g. DatabaseAdmin should not see ServerWideOnly values)

ayende replied:
Where are you doing the filtering?

ppekrol replied:
There was no intention to do any filtering. It depends how we will design the UI.

ayende replied:
Let me rephrase, where are you replacing the value with `****`

ppekrol replied:
Done

ayende commented on src/Raven.Server/Config/Attributes/ConfigurationEntryAttribute.cs:
```diff
@@ -12,11 +12,12 @@ namespace Raven.Server.Config.Attributes
             Key = key;
             Order = order;
             SetDefaultValueIfNeeded = setDefaultValueIfNeeded;
+            IsSecured = isSecured;
```
I don't understand the purpose of this? Are we talking about private information? Such as passwords?

ppekrol replied:
It is for the UI, so we can substitute the value with `***********` with 'button' to see the real value

ayende replied:
Why? The arguments, I can understand, but the exec doesn't seems like it should be.

ppekrol replied:
Depends on the angle. I prefer to hide it, because it reveals a path.

ppekrol replied:
Done

ayende commented on src/Raven.Server/Config/Categories/SecurityConfiguration.cs:
```diff
@@ -10,160 +10,161 @@
         [Description("Disable automatic redirection when listening to HTTPS. By default, when using port 443, RavenDB redirects all incoming HTTP traffic on port 80 to HTTPS on port 443.")]
         [DefaultValue(false)]
-        [ConfigurationEntry("Security.DisableHttpsRedirection", ConfigurationEntryScope.ServerWideOnly)]
+        [ConfigurationEntry("Security.DisableHttpsRedirection", ConfigurationEntryScope.ServerWideOnly, isSecured: true)]
```
```suggestion
        [ConfigurationEntry("Security.DisableHttpsRedirection", ConfigurationEntryScope.ServerWideOnly, isSecured: false)]
```

ayende commented on src/Raven.Server/Config/Categories/SecurityConfiguration.cs:
```diff
@@ -10,160 +10,161 @@
         [Description("The path to a folder where RavenDB will store the access audit logs")]
         [DefaultValue(null)]
-        [ConfigurationEntry("Security.AuditLog.FolderPath", ConfigurationEntryScope.ServerWideOnly)]
+        [ConfigurationEntry("Security.AuditLog.FolderPath", ConfigurationEntryScope.ServerWideOnly, isSecured: true)]
```
Not secret

ayende commented on src/Raven.Server/Config/Categories/SecurityConfiguration.cs:
```diff
@@ -10,160 +10,161 @@
-        [ConfigurationEntry("Security.AuditLog.RetentionTimeInHours", ConfigurationEntryScope.ServerWideOnly)]
+        [ConfigurationEntry("Security.AuditLog.RetentionTimeInHrs", ConfigurationEntryScope.ServerWideOnly, isSecured: true)]
+        [ConfigurationEntry("Security.AuditLog.RetentionTimeInHours", ConfigurationEntryScope.ServerWideOnly, isSecured: true)]
         public TimeSetting AuditLogRetentionTime { get; set; }
```
Not secret

ayende commented on src/Raven.Server/Config/Categories/SecurityConfiguration.cs:
```diff
@@ -10,160 +10,161 @@
         [MinValue(256)]
         [SizeUnit(SizeUnit.Megabytes)]
-        [ConfigurationEntry("Security.AuditLog.RetentionSizeInMb", ConfigurationEntryScope.ServerWideOnly)]
+        [ConfigurationEntry("Security.AuditLog.RetentionSizeInMb", ConfigurationEntryScope.ServerWideOnly, isSecured: true)]
```
A lot of those values are not secrets.
Basically, just passwords should be hidden

ppekrol committed:
RavenDB-5042
- when configuration is secured we do not want to display value at all
- removed couple isSecured configurations

ppekrol committed:
RavenDB-5042 added the ability to filter

ppekrol commented:
@ayende I've applied your comments. Removed isSecured from most of the configuration entries. Left few. Could you look at them?

ppekrol committed:
RavenDB-5042 added /admin/configuration/settings endpoint just for server values

ayende commented on src/Raven.Server/Config/Categories/SecurityConfiguration.cs:
```diff
@@ -104,12 +105,12 @@ public class SecurityConfiguration : ConfigurationCategory
         [Description("The E-mail address associated with the Let's Encrypt certificate. Used for renewal requests.")]
         [DefaultValue(null)]
-        [ConfigurationEntry("Security.Certificate.LetsEncrypt.Email", ConfigurationEntryScope.ServerWideOnly)]
+        [ConfigurationEntry("Security.Certificate.LetsEncrypt.Email", ConfigurationEntryScope.ServerWideOnly, isSecured: true)]
```
Not a secret, probably

ayende commented on src/Raven.Server/Config/Categories/SecurityConfiguration.cs:
```diff
@@ -104,12 +105,12 @@ public class SecurityConfiguration : ConfigurationCategory
         [Description("The path of the (256-bit) Master Key. If specified, RavenDB will use this key to protect secrets.")]
         [DefaultValue(null)]
-        [ConfigurationEntry("Security.MasterKey.Path", ConfigurationEntryScope.ServerWideOnly)]
+        [ConfigurationEntry("Security.MasterKey.Path", ConfigurationEntryScope.ServerWideOnly, isSecured: true)]
```
Why is that a secret? It convey no additional information

ayende commented on src/Raven.Server/Config/Categories/SecurityConfiguration.cs:
```diff
@@ -135,12 +136,12 @@ public class SecurityConfiguration : ConfigurationCategory
         [DefaultValue(null)]
-        [ConfigurationEntry("Security.WellKnownIssuerHashes.Admin", ConfigurationEntryScope.ServerWideOnly)]
+        [ConfigurationEntry("Security.WellKnownIssuerHashes.Admin", ConfigurationEntryScope.ServerWideOnly, isSecured: true)]
         public string[] WellKnownIssuerHashes { get; set; }
```
Not a secret (and we report that to user anyway)

ayende commented on src/Raven.Server/Config/Categories/SecurityConfiguration.cs:
```diff
@@ -135,12 +136,12 @@ public class SecurityConfiguration : ConfigurationCategory
         [DefaultValue(null)]
-        [ConfigurationEntry("Security.WellKnownCertificates.Admin", ConfigurationEntryScope.ServerWideOnly)]
+        [ConfigurationEntry("Security.WellKnownCertificates.Admin", ConfigurationEntryScope.ServerWideOnly, isSecured: true)]
         public string[] WellKnownAdminCertificates { get; set; }
```
Not a secret (and we report that to user anyway)

ppekrol committed:
RavenDB-5042 removed isSecured from few SecurityConfiguration entries
